discretize/CylMesh.py

```diff
@@ -123,72 +227,531 @@ def vectorCCx(self):
+            return np.kron(
+                    np.ones_like(self.vectorNz), pi*(
+                        self.vectorNx**2 -
+                        np.r_[0, self.vectorNx[:-1]]**2
+                    )
+            )
+        return np.kron(
+            np.ones(self._ntNz), np.kron(
+                self.hy,
+                0.5 * (self.vectorNx[1:]**2 - self.vectorNx[:-1]**2)
+            )
+        )
+
+
+    @property
+    def areaFz(self):
+        if getattr(self, '_areaFz', None) is None:
+            areaFzFull = self._areaFzFull
             if self.isSymmetric:
-                self._edge = 2*pi*self.gridN[:, 0]
+                self._areaFz = areaFzFull
             else:
-                raise NotImplementedError('edges not yet implemented for 3D '
-                                          'cyl mesh')
-        return self._edge
+                self._areaFz = (
+                    self._deflationMatrix('Fz', withHanging=False) *
+                    areaFzFull
+                )
+        return self._areaFz
+
+    @property
+    def _areaFull(self):
+        return np.r_[self._areaFxFull, self._areaFyFull, self._areaFzFull]
 
     @property
     def area(self):
         """Face areas"""
-        if getattr(self, '_area', None) is None:
-            if self.nCy > 1:
-                raise NotImplementedError('area not yet implemented for 3D '
-                                          'cyl mesh')
-            areaR = np.kron(self.hz, 2*pi*self.vectorNx)
-            areaZ = np.kron(np.ones_like(self.vectorNz), pi*(self.vectorNx**2 -
-                            np.r_[0, self.vectorNx[:-1]]**2))
-            self._area = np.r_[areaR, areaZ]
-        return self._area
+        # if getattr(self, '_area', None) is None:
+        if self.isSymmetric is True:
+            return np.r_[self.areaFx, self.areaFz]
+        else:
+            return np.r_[self.areaFx, self.areaFy, self.areaFz]
 
     @property
     def vol(self):
         """Volume of each cell"""
         if getattr(self, '_vol', None) is None:
-            if self.nCy > 1:
-                raise NotImplementedError('vol not yet implemented for 3D '
-                                          'cyl mesh')
-            az = pi*(self.vectorNx**2 - np.r_[0, self.vectorNx[:-1]]**2)
-            self._vol = np.kron(self.hz, az)
+            if self.isSymmetric:
+                az = pi*(self.vectorNx**2 - np.r_[0, self.vectorNx[:-1]]**2)
+                self._vol = np.kron(self.hz, az)
+            else:
+                self._vol = np.kron(
+                    self.hz, np.kron(
+                        self.hy,
+                        0.5 * (self.vectorNx[1:]**2 - self.vectorNx[:-1]**2)
+                    )
+                )
         return self._vol
 
     ####################################################
+    # Active and Hanging Edges and Faces
+    ####################################################
+
+    @property
+    def _ishangingFx(self):
```

Comment:
I am not sure if krons of bools is the most elegant way to do this, but it was the clearest in my head. I would appreciate hearing thoughts!

Reply:
ha, is that horrible??